Make memory hooks work with static libcudart, by using bistro hooks for cuda driver API

yosefe opened the pull request:
# Why
Support Cuda memory allocate/release hooks also for applications which are linked to libcudart statically, or use directly the driver API.
Fixed #3210

# How
Use binary instrumentation (bistro) hooks for Cuda Driver APIs, which work regardless of how a function is called (from .plt or by function pointer). 
- Before this PR, if an application was statically linked to libcudart, we could not install hooks on Cuda memory events: cuda runtime calls (e.g cudaFree) are not going through .plt, and libcudart is calling driver API (e.g cuMemFree) by function pointer.
- After this PR, the calls from cudart to cuda driver API can be intercepter.
 
If any other library is intercepting the same function by .plt hooks, it can work alongside with bistro hooks. However, if the other library is also using binary patching, in will probably be incompatible with the UCX hooks.

Binary instrumentation hook method:
1. Copy minimal number of bytes of target function prologue to a newly allocated buffer with execute permissions.
2. After these code bytes, place an indirect jump instruction to the rest of the target function code.
     > If there are not enough position independent instructions, such as `push <reg>` or `sub <imm>, <reg>`, to accommodate 
     > for the indirect jump (3), this method will fail.
2. Place an indirect jump in the target function start to the UCM replacement
3. UCM replacement will treat the executable buffer (1) as the "original" function which should be used from the replacement (2)
```
memory allocate/release
in the application
      +
      |
      |       target function
      |       (e.g cuMemAlloc)               UCM replacement
      +----->-----------------+              (e.g ucm_cuMemAlloc)
             |jmp to UCM hook +-------------->+-----------+
             |                | (3)           | dispatch  +-------------> [ rcache ]
             |                |               | UCM event |
             +----------------+               |           |
             | unmodified     +<---+          +-----------+
             | code of target |    |          | call orig | (4)
             | function       |    |      +---+ function  |
             |                |    |      |   +-----------+
             |                |    |      |
             |                |    |      |   executable buffer
             |                |    |      |   (e.g ucm_orig_cuMemAlloc)
             |                |    |      +-->+-----------+
             |                |    |          | copy of   | (1)
             |                |    |          | function  |
             |                |    |          | prologue  |
             |                |    |          +-----------+
             |                |    +----------+ jump back | (2)
             +----------------+               +-----------+

```

hoopoepg commented on src/ucm/bistro/bistro_x86_64.c:
```diff
@@ -14,19 +14,153 @@
+        }
+
+        /* check the opcode */
+        if (((rex == 0) || rex == 0x41) && ((opcode & 0xF0) == 0x50)) {
```
```((rex == 0) || rex == 0x41)``` is this correct?

yosefe replied:
rex == 0 is pushing IA32 register and rex==0x41 is pushing rN register

hoopoepg commented on src/ucm/bistro/bistro_x86_64.c:
```diff
@@ -14,19 +14,153 @@
+            }
+        }
+
+        /* unsupported instruction - bail */
```
bail? not fail?

yosefe replied:
yes, "bail" means exit

hoopoepg commented on src/ucm/bistro/bistro_x86_64.c:
```diff
@@ -14,19 +14,153 @@
+
+    pthread_mutex_lock(&mutex);
+
+    if (page == MAP_FAILED) {
```
this condition is always true

hoopoepg commented on src/ucm/bistro/bistro_x86_64.c:
```diff
@@ -14,19 +14,153 @@
+        }
+    }
+
+    if (size + offset > ucm_get_page_size()) {
```
offset is 0 here... and wrap ```size + offset``` by ()

hoopoepg commented on src/ucm/bistro/bistro_x86_64.c:
```diff
@@ -14,19 +14,153 @@
+
+    /* allocate space in the page */
+    result  = UCS_PTR_BYTE_OFFSET(page, offset);
+    offset += size;
```
it seems body of this function was in loop :)

yosefe replied:
forgot to add 'static'

yosefe commented:
bot:pipe:retest

yosefe commented:
/azp run

yosefe commented:
bot:pipe:retest

yosefe commented:
/azp run

yosefe commented:
@Akshay-Venkatesh @bureddy can you pls take a look?

yosefe commented:
@bureddy fixed

Akshay-Venkatesh commented:
> 
> 
> @Akshay-Venkatesh @bureddy can you pls take a look?

Some high-level questions/requests:
 - Can you please explain how/why this solves the static linking issue? "How" section in the PR seems to describe the low-level mechanism of binary instrumentation. Specifying current limitations with cudart static linking (such as use of function pointers in cudart), how current hooks in ucm/cuda fails to intercept when cudart static is used, and how this PR addresses issues could be useful for future reference. 
 - Should we document any limitations with current PR (such as memory allocated before ucp_init)?
 - Does this PR address the case when there are other libraries intercept cu* functions?

yosefe commented:
> * Can you please explain how/why this solves the static linking issue? "How" section in the PR seems to describe the low-level mechanism of binary instrumentation. Specifying current limitations with cudart static linking (such as use of function pointers in cudart), how current hooks in ucm/cuda fails to intercept when cudart static is used, and how this PR addresses issues could be useful for future reference.

Added

> * Should we document any limitations with current PR (such as memory allocated before ucp_init)?

Memory before ucp_init is not a problem: we set suspicious regions to UNKNOWN and force detection during runtime

> * Does this PR address the case when there are other libraries intercept cu* functions?

Added note: if other library is hooking .plt it would work, if using bistro - no

bureddy commented on src/ucm/cuda/cudamem.c:
```diff
@@ -6,392 +6,263 @@
-cudaError_t ucm_cudaHostGetDevicePointer(void **pDevice, void *pHost, unsigned int flags)
-{
-    cudaError_t ret;
+        if (ucm_mmap_hook_mode() == UCM_MMAP_HOOK_BISTRO) {
```
@yosefe  should we detach cuda hooks from ucm_mmap_hook_mode() flag? may be we have a case where we need to to disable malloc bistro hooks but not cuda hooks ( for ex: on summit with spectrum mpi).
how about controlling cuda hooks with new flag "ucm_cuda_hook_mode: and "remove ucm_global_opts.enable_cuda_reloc" ?

bureddy commented on src/ucm/cuda/cudamem.c:
```diff
@@ -261,7 +266,7 @@ static ucs_status_t ucm_cudamem_install(int events)
 
-    if (!ucm_global_opts.enable_cuda_reloc) {
+    if (ucm_cuda_hook_mode() == UCM_MMAP_HOOK_NONE) {
         ucm_debug("cudamem relocations are disabled by configuration");
```
s/relocations/hooks?

bureddy commented on src/ucs/config/ucm_opts.c:
```diff
@@ -68,10 +69,21 @@ static ucs_config_field_t ucm_global_config_table[] = {
+   " reloc  - Use ELF relocation table to set hooks.\n"
+#if UCM_BISTRO_HOOKS
+   " bistro - Use binary instrumentation to set hooks. In this mode, calls to\n"
+   "          statically a statically Cuda runtime can be hooked on the the\n"
```
seems wording is not clear

bureddy commented on src/ucm/cuda/cudamem.c:
```diff
@@ -402,17 +278,23 @@ static ucs_status_t ucm_cudamem_install(int events)
+        status = ucm_cuda_install_hooks(ucm_cuda_runtime_funcs, &used_reloc);
+        if (status == UCS_OK) {
+            ucm_debug("installed cuda memory %s hooks on runtime API",
+                      used_reloc ? "reloc" : "bistro");
```
seems "ucm_cuda_install_hooks" can do bistro for some funcs if it is possible and reloc for others. is this debug msg accurate?
